linkerd/core/src/main/scala/io/buoyant/linkerd/TracePropagator.scala

```diff
@@ -0,0 +1,42 @@
+package io.buoyant.linkerd
+
+import com.twitter.finagle.buoyant.{Sampler, TraceInitializer}
+import com.twitter.finagle.tracing.{Trace, TraceId, Tracer}
+
+trait TracePropagator[Req] {
+
+  /**
+   * Read the trace id from the request, if it has one.
+   */
+  def traceId(req: Req): Option[TraceId]
+
+  /**
+   * Return a sampler which decides if the given request should be sampled, based on properties
+   * of the request.  If None is returned, the decision of whether to sample the request is deferred
+   * to the tracer.
+   */
+  def sampler(req: Req): Option[Sampler]
```

Comment:
Presumably, we just hand the request right back to the sampler. Can this API be simplified?

Reply:
Not quite.  Sampler takes a Long as input, not a request.  We feed the request's trace id into the sampler.  I think we want to ensure Sampler is used here rather than just returning Option[Boolean] to ensure that sampling is based on the trace id and is properly salted.

Reply:
Why do we need to pass a request to the interface? so that it can modify headers? should the method name be updated to indicate it's usually side-effecty?